src/garage/_dtypes.py

```diff
@@ -428,22 +431,29 @@ def next_observations(self):
             ]))
 
 
+class StepType(enum.IntEnum):
+    """Defines the status of a `TimeStep` within a sequence."""
+    # Denotes the first `TimeStep` in a sequence.
+    FIRST = 0
+    # Denotes any `TimeStep` in a sequence that is not FIRST or LAST.
+    MID = 1
+    # Denotes the last `TimeStep` in a sequence.
+    LAST = 2
```

Comment:
If you're going to have `TIMEOUT` then this should be named `TERMINALS` to distinguish it from `TIMEOUT`.

`TIMEOUT` is also a `LAST` step. This could actually accidentally make the API harder to use. I'm not sure it's bad (it forces you to make sure you're getting what you asked for), but it might also introduce a sharp edge.

Imagine I want to retrieve all the last observations. Without `TIMEOUT` I would do

```python
last_obs = tsb.next_observations[tsb.step_types == StepType.TERMINAL]
```

With `TIMEOUT` you need
```python
last_obs = tsb.next_observations[tsb.step_types == StepType.TERMINAL | tsb.step_types == StepType.TIMEOUT]
```

**Ways of dealing with this**
* Do nothing

* Use a different field for indicating timeout
    ```py
    last_obs = tsb.next_observations[tsb.step_types == StepType.LAST]  # includes TERMINAL and TIMEOUT
    # timeout_obs = last_obs[tsb.timeouts]  <-- oof, you actually can't do this
    timeout_obs = tsb.next_observations[tsb.timeouts]  # <-- okay this should work fine  # just timeouts
    ```

* Leave it as is, but add a helper function

    ```py
    def get_last_obs(self):
        return self.next_observations[self.step_types == StepType.TERMINAL | self.step_types == StepType.TIMEOUT]

* Switch up the Enum to make TERMINAL the special case

    ```python
    last_obs = tsb.next_observations[tsb.step_types == StepType.LAST]  # includes TERMINAL and TIMEOUT
    terminal_obs = tsb.next_observations[tsb.terminals]  # just terminals
    ```

**How to decide:**
Try refactoring functions like these using different options for this design, and see which one leads to more intuitive/less error-prone code. Note that in the off-policy algorithms, terminals means specifically *not* a timeout step, whereas `paths_to_tensors` (for the on-policy algorithms) doesn't care.

https://github.com/rlworkgroup/garage/blob/f5e8cbc21c5a78e2c2996e256f0f2bd5b6930f90/src/garage/torch/algos/sac.py#L370
https://github.com/rlworkgroup/garage/blob/f5e8cbc21c5a78e2c2996e256f0f2bd5b6930f90/src/garage/tf/algos/ddpg.py#L370
https://github.com/rlworkgroup/garage/blob/f5e8cbc21c5a78e2c2996e256f0f2bd5b6930f90/src/garage/tf/_functions.py#L81
https://github.com/rlworkgroup/garage/blob/f5e8cbc21c5a78e2c2996e256f0f2bd5b6930f90/src/garage/tf/algos/dqn.py#L154

Reply:
Thanks for this very helpful comment. Since we are certain that `reset()` and `step()` should return `TimeStep`, I will create a new PR just for TimeStep as well as the implementations, and see how different choices work. 

This PR now only introduces src/garage/environment.py

Reply:
Did you do this exercise? What were the results? I think it's a real design concern.

Reply:
See https://github.com/rlworkgroup/garage/pull/1786 for refactoring implementation.